File: javascript/src/notify/notify.test.ts
// Tests for the Slack/Teams notifiers — stubs global fetch to capture posted bodies.
import { afterEach, describe, expect, it, vi } from 'vitest';
import {
  type Message,
  newNotifier,
  SlackNotifier,
  TeamsNotifier,
  teamsCard,
} from './notify';

const URL = 'https://hook.example/webhook';

function stubFetch(status: number, text = ''): ReturnType<typeof vi.fn> {
  const fn = vi.fn(async () => new Response(text, { status }));
  vi.stubGlobal('fetch', fn);
  return fn;
}

function lastBody(fn: ReturnType<typeof vi.fn>): any {
  const init = fn.mock.calls.at(-1)![1] as RequestInit;
  return JSON.parse(init.body as string);
}

afterEach(() => {
  vi.unstubAllGlobals();
});

describe('notify', () => {
  it('posts a Slack mrkdwn body', async () => {
    const fn = stubFetch(200);
    const m: Message = { title: 'Digest', text: '3 commits', link: 'https://x/pr/1' };
    await new SlackNotifier(URL).notify(m);

    expect(fn).toHaveBeenCalled();
    expect(lastBody(fn).text).toBe('*Digest*\n3 commits\n<https://x/pr/1>');
  });

  it('posts a Teams Adaptive Card', async () => {
    const fn = stubFetch(200);
    await new TeamsNotifier(URL).notify({ title: 'Result', text: 'fixed', link: 'https://x/pr/2' });

    expect(fn).toHaveBeenCalled();
    const payload = lastBody(fn);
    expect(payload.type).toBe('message');
    expect(payload.attachments).toHaveLength(1);
    const att = payload.attachments[0];
    expect(att.contentType).toBe('application/vnd.microsoft.card.adaptive');
    expect(att.content.type).toBe('AdaptiveCard');
    expect(att.content.actions).toBeDefined();
  });

  it('treats a non-2xx response as an error', async () => {
    stubFetch(500, 'boom');
    await expect(new SlackNotifier(URL).notify({ text: 'x' })).rejects.toThrow();
  });

  it('picks the implementation by provider', () => {
    expect(newNotifier('slack', 'https://hook', '')).toBeInstanceOf(SlackNotifier);
    expect(newNotifier('teams', '', 'https://hook')).toBeInstanceOf(TeamsNotifier);
    expect(() => newNotifier('slack', '', '')).toThrow();
    expect(() => newNotifier('discord', 'a', 'b')).toThrow();
  });

  it('omits Teams actions without a link', () => {
    const card = teamsCard({ title: 't', text: 'b' });
    const content = (card.attachments as any)[0].content;
    expect(content.actions).toBeUndefined();
  });
});

File: go/internal/notify/notify_test.go
package notify

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	"automation-agent/internal/useragent"
)

// captureServer records the body of the next POST it receives.
func captureServer(t *testing.T, status int) (*httptest.Server, *[]byte) {
	t.Helper()
	var got []byte
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		b, _ := io.ReadAll(r.Body)
		got = b
		w.WriteHeader(status)
	}))
	t.Cleanup(srv.Close)
	return srv, &got
}

func TestSlackNotify(t *testing.T) {
	srv, body := captureServer(t, http.StatusOK)
	n := newSlack(srv.URL)

	if err := n.Notify(context.Background(), Message{Title: "Digest", Text: "3 commits", Link: "https://x/pr/1"}); err != nil {
		t.Fatalf("Notify: %v", err)
	}
	var payload map[string]string
	if err := json.Unmarshal(*body, &payload); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	want := "*Digest*\n3 commits\n<https://x/pr/1>"
	if payload["text"] != want {
		t.Errorf("text = %q, want %q", payload["text"], want)
	}
}

func TestTeamsNotify(t *testing.T) {
	srv, body := captureServer(t, http.StatusOK)
	n := newTeams(srv.URL)

	if err := n.Notify(context.Background(), Message{Title: "Result", Text: "fixed", Link: "https://x/pr/2"}); err != nil {
		t.Fatalf("Notify: %v", err)
	}
	var payload map[string]any
	if err := json.Unmarshal(*body, &payload); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if payload["type"] != "message" {
		t.Errorf("type = %v, want message", payload["type"])
	}
	atts, ok := payload["attachments"].([]any)
	if !ok || len(atts) != 1 {
		t.Fatalf("attachments = %v", payload["attachments"])
	}
	att := atts[0].(map[string]any)
	if att["contentType"] != "application/vnd.microsoft.card.adaptive" {
		t.Errorf("contentType = %v", att["contentType"])
	}
	content := att["content"].(map[string]any)
	if content["type"] != "AdaptiveCard" {
		t.Errorf("content type = %v", content["type"])
	}
	if _, hasActions := content["actions"]; !hasActions {
		t.Error("expected actions for a message with a link")
	}
}

func TestNon2xxIsError(t *testing.T) {
	srv, _ := captureServer(t, http.StatusInternalServerError)
	if err := newSlack(srv.URL).Notify(context.Background(), Message{Text: "x"}); err == nil {
		t.Fatal("expected error on 500")
	}
}

func TestNewFactory(t *testing.T) {
	if _, err := New("slack", "https://hook", ""); err != nil {
		t.Errorf("slack: %v", err)
	}
	if _, err := New("teams", "", "https://hook"); err != nil {
		t.Errorf("teams: %v", err)
	}
	if _, err := New("slack", "", ""); err == nil {
		t.Error("slack without URL should error")
	}
	if _, err := New("discord", "a", "b"); err == nil {
		t.Error("unknown provider should error")
	}
}

func TestTeamsCardOmitsActionsWithoutLink(t *testing.T) {
	card := teamsCard(Message{Title: "t", Text: "b"})
	content := card["attachments"].([]map[string]any)[0]["content"].(map[string]any)
	if _, has := content["actions"]; has {
		t.Error("no link -> no actions")
	}
}

// Every outbound request identifies this service. Asserted at the wire rather than on the
// header-setting call, because the value a Slack or Teams admin sees in their logs is the
// only thing that matters here.
func TestNotificationsIdentifyTheService(t *testing.T) {
	var ua string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		ua = r.Header.Get("User-Agent")
		w.WriteHeader(http.StatusOK)
	}))
	defer srv.Close()

	if err := postJSON(context.Background(), defaultClient(), srv.URL, map[string]string{"text": "x"}); err != nil {
		t.Fatalf("postJSON: %v", err)
	}
	if want := useragent.String(); ua != want {
		t.Errorf("webhook saw User-Agent %q, want %q", ua, want)
	}
}
